{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a8e68a8",
   "metadata": {},
   "source": [
    "# New QA tests\n",
    "\n",
    "We are going to make new QA tests, by reconstructing bigger trees."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a4e6d266",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../')\n",
    "from qa_functions import *\n",
    "from colorama import Fore\n",
    "import time\n",
    "import networkx as nx\n",
    "import random\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e207483",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_names = os.listdir('alignments')\n",
    "\n",
    "# Remove the folder test\n",
    "file_names = file_names[:-1]\n",
    "\n",
    "for folder in file_names:\n",
    "    if int(folder) > 31:\n",
    "        files = os.listdir(f'alignments/{folder}')\n",
    "        for file in files:\n",
    "            \n",
    "            print(f'alignments/{folder}/{file}')\n",
    "            # Load the sequences and create simmilarity matrix\n",
    "            sequences,labels = load_sequences(f'./alignments/{folder}/{file}',format='fasta')\n",
    "            distance_matrix = compute_distance_matrix(sequences)\n",
    "\n",
    "            # Save the distance matrix\n",
    "            file_ext = re.search(r'_([0-9]+)\\.',file).group(1)\n",
    "            matrix_file = new_file = f'matrix_{file_ext}'\n",
    "            np.save(f'./matrices/{folder}/{matrix_file}',distance_matrix)\n",
    "            \n",
    "            # Begin NMcutQA\n",
    "            timer = Timer(0.0)\n",
    "            tree_qa = phylo_tree(distance_matrix,timer=timer)\n",
    "            \n",
    "            with open('timer.csv','a') as fp:\n",
    "                fp.write(f'{folder},{file},{timer.value}\\n')\n",
    "            \n",
    "            new_file = f'qa_tree_{file_ext}.newick'\n",
    "            tree_qa.create_newick_file(f'./trees/{folder}/{new_file}')\n",
    "            print(f'Finished {file} in {timer.value}ms')\n",
    "\n",
    "            # Erase memory\n",
    "            del sequences\n",
    "            del labels\n",
    "            del distance_matrix\n",
    "            del tree_qa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "122c330a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alignments/43/assembled_sequences_49038.fasta\n",
      "Finished assembled_sequences_49038.fasta in 2446.8066400000002ms\n"
     ]
    }
   ],
   "source": [
    "file = 'assembled_sequences_49038.fasta'\n",
    "folder = '43'\n",
    "\n",
    "print(f'alignments/{folder}/{file}')\n",
    "# Load the sequences and create simmilarity matrix\n",
    "sequences,labels = load_sequences(f'../alignments/{folder}/{file}',format='fasta')\n",
    "distance_matrix = compute_distance_matrix(sequences)\n",
    "\n",
    "# Save the distance matrix\n",
    "file_ext = re.search(r'_([0-9]+)\\.',file).group(1)\n",
    "matrix_file = new_file = f'matrix_{file_ext}'\n",
    "np.save(f'../matrices/{folder}/{matrix_file}',distance_matrix)\n",
    "\n",
    "# Begin NMcutQA\n",
    "timer = Timer(0.0)\n",
    "tree_qa = phylo_tree(distance_matrix,timer=timer)\n",
    "\n",
    "with open('../metrics/timer.csv','a') as fp:\n",
    "    fp.write(f'{folder},{file},{timer.value}\\n')\n",
    "\n",
    "new_file = f'qa_tree_{file_ext}.newick'\n",
    "tree_qa.create_newick_file(f'../trees/{folder}/{new_file}')\n",
    "print(f'Finished {file} in {timer.value}ms')\n",
    "\n",
    "# Erase memory\n",
    "del sequences\n",
    "del labels\n",
    "del distance_matrix\n",
    "del tree_qa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9b2522a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Robinson-Foulds distance bewteen the Ncut with QA tree and the Neighbor-Joining tree is \u001b[1m25.0\u001b[22m\n"
     ]
    }
   ],
   "source": [
    "with open(f'../trees/{folder}/{new_file}','r') as file:\n",
    "    qa_tree = file.read()\n",
    "\n",
    "with open(f'../trees/{folder}/tree_best_{file_ext}.newick','r') as file:\n",
    "    biotree = file.read()\n",
    "biotree = re.sub(r'taxon([0-9]+)',lambda match: str(int(match.group(1)) - 1),biotree)\n",
    "biotree = re.sub(r':[0-9]\\.[0-9]+(e-[0-9]+)*',r'',biotree)\n",
    "\n",
    "dist = treecmp(qa_tree,biotree)\n",
    "\n",
    "print(f'The Robinson-Foulds distance bewteen the Ncut with QA tree and the Neighbor-Joining tree is \\033[1m{dist}\\033[22m')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baac2c1c",
   "metadata": {},
   "source": [
    "We have generated alignments for trees from size 100 to 150. Remember, the expected number of subproblems is given by the expression:\n",
    "$$\\mu(x)=0.35x\\cdot \\log_2{10.82x}.$$\n",
    "\n",
    "Given that a problem usually takes around 3 seconds to be enter and run in the remote QPU, the expected total time for the execution will be:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6c8e252c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The full execution will take approximately 9.87 hours\n"
     ]
    }
   ],
   "source": [
    "def f(x):\n",
    "    return 0.35*x*np.log2(10.82*x)\n",
    "\n",
    "x = np.arange(100,151,2)\n",
    "print(f'The full execution will take approximately {round(np.sum(f(x)*3/60)/60,3)} hours')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c7fc61f",
   "metadata": {},
   "source": [
    "We can run the program here, but it should be run in CESGA or similar."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c400d507",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alignments/100/assembled_sequences_53321.fasta\n",
      "Alignment file generated, generating the phylogenetic tree...\n"
     ]
    }
   ],
   "source": [
    "file_names = os.listdir('../alignments')\n",
    "\n",
    "# Remove the folder test\n",
    "file_names = file_names[:-1]\n",
    "\n",
    "for folder in file_names:\n",
    "    if int(folder) > 99 and int(folder) < 125:\n",
    "        files = os.listdir(f'../alignments/{folder}')\n",
    "        for file in files:\n",
    "            \n",
    "            print(f'alignments/{folder}/{file}')\n",
    "            # Load the sequences and create simmilarity matrix\n",
    "            sequences,labels = load_sequences(f'../alignments/{folder}/{file}',format='fasta')\n",
    "            distance_matrix = compute_distance_matrix(sequences)\n",
    "\n",
    "            # Save the distance matrix\n",
    "            file_ext = re.search(r'_([0-9]+)\\.',file).group(1)\n",
    "            matrix_file = new_file = f'matrix_{file_ext}'\n",
    "            np.save(f'../matrices/{folder}/{matrix_file}',distance_matrix)\n",
    "            print(f'Alignment file generated, generating the phylogenetic tree...')\n",
    "            # Begin NMcutQA\n",
    "            timer = Timer(0.0)\n",
    "            tree_qa = phylo_tree(distance_matrix,timer=timer)\n",
    "\n",
    "            with open('../metrics/timer.csv','a') as fp:\n",
    "                fp.write(f'{folder},{file},{timer.value}\\n')\n",
    "\n",
    "            new_file = f'qa_tree_{file_ext}.newick'\n",
    "            tree_qa.create_newick_file(f'../trees/{folder}/{new_file}')\n",
    "            print(f'Finished {file} in {timer.value}ms\\n---------------------------------------------------------\\n')\n",
    "\n",
    "            # Erase memory\n",
    "            del sequences\n",
    "            del labels\n",
    "            del distance_matrix\n",
    "            del tree_qa"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
